tasks/ci.rake

```diff
@@ -243,13 +256,37 @@ end
 def beaker_suite(type)
   beaker(:init, '--hosts', ENV['HOSTS'], '--options-file', "config/#{String(type)}/options.rb")
   beaker(:provision)
-  beaker(:exec, 'pre-suite', '--preserve-state', '--pre-suite', pre_suites(type))
-  beaker(:exec, 'pre-suite', '--preserve-state')
-  beaker(:exec, ENV['TESTS'])
-  beaker(:exec, 'post-suite')
+  begin
+    beaker(:exec, 'pre-suite', '--preserve-state', '--pre-suite', pre_suites(type))
+    beaker(:exec, 'pre-suite', '--preserve-state')
+    beaker(:exec, ENV['TESTS'])
+    beaker(:exec, 'post-suite')
+  ensure
+    preserve_hosts = ENV['OPTIONS'].include?('--preserve-hosts=always') if ENV['OPTIONS']
+    beaker(:destroy) unless preserve_hosts
+  end
+end
+
+def beaker_suite_retry(type)
+  beaker(:init, '--hosts', ENV['HOSTS'], '--options-file', "config/#{String(type)}/options.rb")
+  beaker(:provision)
+  begin
+    beaker(:exec, 'pre-suite', '--preserve-state', '--pre-suite', pre_suites(type))
+    beaker(:exec, 'pre-suite', '--preserve-state')
+
+    beaker(:exec, ENV['TESTS'], '--save-results', '/tmp/results.json')
+  rescue RuntimeError
+    tests_to_rerun = get_tests_from('/tmp/results.json')
+    beaker(:exec, tests_to_rerun.map { |str| "#{str}" }.join(',') ) if tests_to_rerun.length > 0
+  ensure
+    beaker(:exec, 'post-suite')
+    preserve_hosts = ENV['OPTIONS'].include?('--preserve-hosts=always') if ENV['OPTIONS']
+    beaker(:destroy) unless preserve_hosts
+  end
+end
 
-  preserve_hosts = ENV['OPTIONS'].include?('--preserve-hosts=always') if ENV['OPTIONS']
-  beaker(:destroy) unless preserve_hosts
+def get_tests_from(results_file)
```

Comment:
Not sure it's worth adding this method to wrap the JSON load

Reply:
Refactored.